core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala

```diff
@@ -0,0 +1,462 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import java.io.File
+import java.net.ConnectException
+import java.nio.ByteBuffer
+import java.util.concurrent.ExecutorService
+
+import scala.collection.mutable.{ArrayBuffer, HashMap, HashSet, Queue}
+
+import com.google.common.base.Throwables
+
+import org.apache.spark.{ShuffleDependency, SparkConf, SparkEnv}
+import org.apache.spark.annotation.Since
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.network.buffer.{FileSegmentManagedBuffer, ManagedBuffer, NioManagedBuffer}
+import org.apache.spark.network.netty.SparkTransportConf
+import org.apache.spark.network.shuffle.BlockFetchingListener
+import org.apache.spark.network.shuffle.ErrorHandler.BlockPushErrorHandler
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.shuffle.ShuffleBlockPusher._
+import org.apache.spark.storage.{BlockId, BlockManagerId, ShufflePushBlockId}
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+/**
+ * Used for pushing shuffle blocks to remote shuffle services when push shuffle is enabled.
+ * When push shuffle is enabled, it is created after the shuffle writer finishes writing the shuffle
+ * file and initiates the block push process.
+ *
+ * @param dataFile         mapper generated shuffle data file
+ * @param partitionLengths array of shuffle block size so we can tell shuffle block
+ *                         boundaries within the shuffle file
+ * @param dep              shuffle dependency to get shuffle ID and the location of remote shuffle
+ *                         services to push local shuffle blocks
+ * @param partitionId      map index of the shuffle map task
+ * @param conf             spark configuration
+ */
+@Since("3.1.0")
+private[spark] class ShuffleBlockPusher(
+    dataFile: File,
+    partitionLengths: Array[Long],
+    dep: ShuffleDependency[_, _, _],
+    partitionId: Int,
+    conf: SparkConf) extends Logging {
```

Comment:
Should we make this consistent with `ShuffleBlockFetchIterator`? For `ShuffleBlockFetchIterator` just instantiating it initiates the fetch. I can do the same for `ShuffleBlockPusher`.

Reply:
For fields like `maxBlockSizeToPush`, `maxBlockBatchSize` and etc., I think we can follow `ShuffleBlockFetchIterator`. I think it helps people to understand codes better when they (fetcher & pusher) under the same pattern.

For fields I mentioned here, like `dataFile`, `partitionLengths` and etc., I think we can make them as parameters of `initiateBlockPush()` instead of the constructor's fields. Because they're only used inside `initiateBlockPush()`, so I think it should be enough to pass in as parameters. WDYT?

Reply:
Yes, it makes to sense to have `dataFile`, `partitionLengths`, and the rest which are used only inside `initiateBlockPush` to pass in a parameters to just that method. I will make that change.

For the constructor, I feel it's better to just pass in the spark `conf` object and let the `ShuffleBlockPusher` read the values of the configuration from it which it needs. It needs 5 configuration and passing them explicitly in the constructor would mean everywhere it is instantiated, the code to read these configs will be repeated (though, currently in src it is created at just 1 place). It also leads to mistakes where values of configs are passed in different order.

Reply:
> For the constructor, I feel it's better to just pass in the spark conf object and let the ShuffleBlockPusher read the values of the configuration from it which it needs.

Sounds fine. I actually do not have a strong opinion on this.

Reply:
I have made the change. Please take a look.